typescript/xstate/src/promise.ts

```diff
@@ -0,0 +1,227 @@
+  AnyActorRef,
+  AnyStateMachine,
+  InvokeConfig,
+  NonReducibleUnknown,
+  Snapshot,
+} from "xstate";
+import {AnyActorSystem} from "xstate/dist/declarations/src/system";
+import {RestateActorSystem, SerialisableActorRef, serialiseActorRef, xStateApi} from "./lib";
+import * as restate from "@restatedev/restate-sdk";
+import {TerminalError} from "@restatedev/restate-sdk";
+
+export type PromiseSnapshot<TOutput, TInput> = Snapshot<TOutput> & {
+  input: TInput | undefined;
+  sent: boolean;
+};
+
+const RESTATE_PROMISE_SENT = 'restate.promise.sent';
+const RESTATE_PROMISE_RESOLVE = 'restate.promise.resolve';
+const RESTATE_PROMISE_REJECT = 'restate.promise.reject';
+const XSTATE_STOP = 'xstate.stop';
+
+type PromiseCreator<TOutput, TInput extends NonReducibleUnknown> = ({input, ctx}: {
+  input: TInput;
+  ctx: restate.RpcContext,
+}) => PromiseLike<TOutput>
+
+export type PromiseActorLogic<TOutput, TInput = unknown> = ActorLogic<
+  PromiseSnapshot<TOutput, TInput>,
+  { type: string; [k: string]: unknown },
+  TInput, // input
+  AnyActorSystem
+> & { sentinel: "restate.promise.actor", config: PromiseCreator<TOutput, TInput> };
+
+export type PromiseActorRef<TOutput> = ActorRefFrom<
+  PromiseActorLogic<TOutput, unknown>
+>;
+
+export function fromPromise<TOutput, TInput extends NonReducibleUnknown>(
+  promiseCreator: PromiseCreator<TOutput, TInput>,
+): PromiseActorLogic<TOutput, TInput> {
+  const logic: PromiseActorLogic<TOutput, TInput> = {
+    sentinel: "restate.promise.actor",
+    config: promiseCreator,
+    transition: (state, event) => {
+      if (state.status !== 'active') {
+        return state;
+      }
+
+      switch (event.type) {
+        case RESTATE_PROMISE_SENT: {
+          return {
+            ...state,
+            sent: true,
+          };
+        }
+        case RESTATE_PROMISE_RESOLVE: {
+          const resolvedValue = (event as any).data;
+          return {
+            ...state,
+            status: 'done',
+            output: resolvedValue,
+            input: undefined
+          };
+        }
+        case RESTATE_PROMISE_REJECT:
+          return {
+            ...state,
+            status: 'error',
+            error: (event as any).data,
+            input: undefined
+          };
+        case XSTATE_STOP:
+          return {
+            ...state,
+            status: 'stopped',
+            input: undefined
+          };
+        default:
+          return state;
+      }
+    },
+    start: (state, {self, system}) => {
+      if (state.status !== 'active') {
+        return;
+      }
+
+      if (state.sent) {
+        return;
+      }
+
+      const rs = (system as RestateActorSystem<any>)
+
+      rs.ctx.send(rs.api.promise).invoke({
+        systemName: rs.systemName,
+        self: serialiseActorRef(self),
+        srcs: actorSrc(self),
+        input: state.input
+      });
+
+      // note that we sent off the promise so we don't do it again
+      (system as any)._relay(self, self, {
```

Comment:
This differs slightly from the built in fromPromise actor because i think there we just rely on in-memory state to keep track of the promise having already been run, whereas this needs to be resistant to rehydration, i guess? in general we may have a problem of re-executing some types of actions, I think